Source/Core/VideoCommon/VertexManagerBase.cpp

```diff
@@ -249,3 +249,44 @@ void VertexManager::DoState(PointerWrap& p)
 {
 	g_vertex_manager->vDoState(p);
 }
+
+void VertexManager::CalculateZSlope(u32 stride)
+{
+	float vtx[9];
+	float out[12];
+
+	// Lookup vertices of the last rendered triangle and software-transform them
+	// This allows us to determine the depth slope, which will be used if zfreeze
+	// is enabled in the following flush.
+	for (unsigned int i = 0; i < 3; ++i)
+	{
+		u8* vtx_ptr = s_pCurBufferPointer - stride * (3 - i);
+		vtx[0 + i * 3] = ((float*)vtx_ptr)[0];
+		vtx[1 + i * 3] = ((float*)vtx_ptr)[1];
+		vtx[2 + i * 3] = ((float*)vtx_ptr)[2];
+
+		VertexShaderManager::TransformToClipSpace(&vtx[i * 3], &out[i * 4]);
+
+		// Transform to Screenspace
+		out[0 + i * 4] = out[0 + i * 4] / out[3 + i * 4] * xfmem.viewport.wd + (xfmem.viewport.xOrig - 342);
+		out[1 + i * 4] = out[1 + i * 4] / out[3 + i * 4] * xfmem.viewport.ht + (xfmem.viewport.yOrig - 342);
+		out[2 + i * 4] = out[2 + i * 4] / out[3 + i * 4] * xfmem.viewport.zRange + xfmem.viewport.farZ;
+	}
+
+	float dx31 = out[8] - out[0];
+	float dx12 = out[0] - out[4];
+	float dy12 = out[1] - out[5];
+	float dy31 = out[9] - out[1];
+
+	float DF31 = out[10] - out[2];
+	float DF21 = out[6] - out[2];
+	float a = DF31 * -dy12 - DF21 * dy31;
+	float b = dx31 * DF21 + dx12 * DF31;
+	float c = -dx12 * dy31 - dx31 * -dy12;
+
```

Comment:
Should we be doing a check here for c = 0?  Otherwise we get a divide by 0 and NaNs that I noticed when stepping through the code here for some polies or is there a bigger issue here?
add something like this? ...

if (abs(c) < FLT_EPSILON)
	{
		PixelShaderManager::SetZSlope(0, 0, 1);
		return;
	}